x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java

```diff
@@ -1183,4 +1204,207 @@ public static void assertResponse(Map<String, Object> expected, Map<String, Obje
+        };
+
+        for (String format : contentMap.keySet()) {
+            Response response = runSqlAsTextWithFormat(builder, format);
+
+            assertEquals(contentMap.get(format), responseBody(response));
+
+            assertTrue(hasText(response.getHeader(HEADER_NAME_ASYNC_ID)));
+            assertEquals("false", response.getHeader(HEADER_NAME_ASYNC_PARTIAL));
+            assertEquals("false", response.getHeader(HEADER_NAME_ASYNC_RUNNING));
+        }
+    }
+
+    public void testAsyncTextPaginated() throws IOException, InterruptedException {
+        final Map<String, String> acceptMap = new HashMap<>() {
+            {
+                put("txt", "text/plain");
+                put("csv", "text/csv");
+                put("tsv", "text/tab-separated-values");
+            }
+        };
+        final int fetchSize = randomIntBetween(1, 10);
+        final int fetchCount = randomIntBetween(1, 9);
+        bulkLoadTestData(fetchSize * fetchCount); // NB: product needs to stay below 100, for txt format tests
+
+        String format = randomFrom(acceptMap.keySet());
+        String mode = randomMode();
+        String cursor = null;
+        for (int i = 0; i <= fetchCount; i++) { // the last iteration (the equality in `<=`) checks on no-cursor & no-results
+            // start the query
+            RequestObjectBuilder builder = (hasText(cursor) ? cursor(cursor) : query("SELECT text, number FROM test")).fetchSize(fetchSize)
+                .waitForCompletionTimeout("0d") // don't wait at all
+                .keepOnCompletion(true)
+                .keepAlive("1d") // keep "forever"
+                .mode(mode)
+                .binaryFormat(false); // prevent JDBC mode to (ignore `format` and) enforce CBOR
+            Response response = runSqlAsTextWithFormat(builder, format);
+
+            Character csvDelimiter = ',';
+
+            assertEquals(200, response.getStatusLine().getStatusCode());
+            assertEquals(response.getHeader(HEADER_NAME_ASYNC_PARTIAL), response.getHeader(HEADER_NAME_ASYNC_RUNNING));
+            String asyncId = response.getHeader(HEADER_NAME_ASYNC_ID);
+            assertTrue(hasText(asyncId));
+
+            // it happens though rarely that the whole response comes through despite the given 0-wait
+            if (response.getHeader(HEADER_NAME_ASYNC_PARTIAL).equals("true")) {
+
+                // potentially wait for it to complete
+                boolean pollForCompletion = randomBoolean();
+                if (pollForCompletion) {
+                    Request request = new Request("GET", SQL_ASYNC_STATUS_REST_ENDPOINT + asyncId);
+                    Map<String, Object> asyncStatus = null;
+                    long millis = 1;
+                    for (boolean isRunning = true; isRunning; Thread.sleep(millis *= 2)) {
+                        asyncStatus = toMap(client().performRequest(request), null);
+                        isRunning = (boolean) asyncStatus.get(IS_RUNNING_NAME);
```

Comment:
Is there a danger here of this `for` not existing and running into a test suite timeout?

Reply:
That would require the server so somehow hang with the search task alive (which I guess would also trigger a timeout).